### PR TITLE
fix: Unregister callbacks when the database (and callbacks) are dropped

### PR DESCRIPTION
**Pull request by fl0rek**
Similar to #23 we encountered this problem when porting to rexie 0.6
Old db callbacks aren't being unregistered now, so dropping the db may result in dropped version change closure being called. Add unregistering callbacks, if they were set by us.

### CODE REVIEW

File: tests/database.rs [hunk 1]
```diff
@@ -142,3 +142,21 @@ async fn test_database_delete_object_store() {
...
+async fn test_database_drop_and_reopen() {
+    let factory = Factory::new().unwrap();
+    factory.delete("test").unwrap().await.unwrap();
+
+    let mut open_request = factory.open("test", Some(1)).unwrap();
+    open_request.on_upgrade_needed(|_| ());
+    let mut database = open_request.await.unwrap();
+    database.on_version_change(|event| event.database().expect("database").close());
+
+    drop(database);
```
Comment (devashishdxt) [inline tests/database.rs:156]: I don't think adding a `Drop` implementation for `Database` is the right approach to solve this. If you call `database.close()` before dropping `database`, it should work fine.

Comment (fl0rek) [reply tests/database.rs:156]: In our testing, both dropping and closing [`Rexie`](https://docs.rs/rexie/latest/rexie/struct.Rexie.html) causes subsequent opens to fail.
<details> 
  <summary>Reopen test case which fails </summary>

```Rust
#[wasm_bindgen_test]
async fn pass() {
    Rexie::delete("test").await.unwrap();

    let rexie = Rexie::builder("test")
        .version(1)
        .add_object_store(
            ObjectStore::new("test-store")
            .key_path("id")
            .auto_increment(true)
            .add_index(Index::new("hash-index", "hash").unique(true))
            )
        .build()
        .await
        .expect("first open");

    // drop and close both cause slightly different js exceptions
    //drop(rexie);
    rexie.close();

    gloo_timers::future::sleep(Duration::from_secs(1)).await;

    let rexie = Rexie::builder("test")
        .version(2)
        .add_object_store(
            ObjectStore::new("test-store")
            .key_path("id")
            .auto_increment(true)
            .add_index(Index::new("hash-index", "hash").unique(true))
            )
        .build()
        .await
        .expect("second open");
}
```
</details>

Without this patch, with `close()` I get `JS exception that was thrown: RuntimeError: unreachable executed` and drop causes `Uncaught Error: closure invoked recursively or after being dropped`.

Comment (devashishdxt) [reply tests/database.rs:156]: I added test in `rexie` [here](https://github.com/devashishdxt/rexie/pull/48). Seems to be working.

Comment (fl0rek) [reply tests/database.rs:156]: This is really weird. CI indeed is green, but when I run same tests locally it fails again, both chromium and firefox. I'm still trying few different things, upgrading wasm-pack or switching to nightly didn't work so far.

Comment (devashishdxt) [reply tests/database.rs:156]: Can you try running `cargo update`?

**Comment by fl0rek** [comment]
@devashishdxt I'm going to close this one as the `close()` workaround works for us. Thanks for the help!

I'm still worried this is a bit of a footgun, maybe there should be note about dropping `Database` and `Rexie` in the docs? I can open the PRs if you don't mind.

**Comment by devashishdxt** [comment]
Thanks @fl0rek. Please feel free to open a PR.

**Pull request closed without merging**